### ISSUE TITLE
erroneous structural cycle detection

### ISSUE THREAD

**Issue by cueckoo** [cue-lang/cue#990]
_Originally opened by @kumorid in https://github.com/cuelang/cue/issues/990_

### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
cue version 0.4.0-beta.2 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?
yes

### What did you do?
I am trying to transform a potentially complex recursive structure into a flattened (one-level) dictionary

The file I am using as a simplified example is

<pre>
-- test.cue --
// A simple discriminated element
#AC: {
  k: "simple"
  d: #C
}

// A recursive discriminated element
#AS: {
  k: "complex"
  d: #S
}

#R: {
  a: #AC | #AS
}

// The structure of the recursive element
#S: {
  n: string
  r: [string]: #R
}

// The structure of the simple element
#C: {
  n: string
}

// Problematic traversal.
#sub: {
  $p=#p: #AS
  dict: [string]: _

  for rn, rd in $p.d.r if rd.a.k == "complex" {
    let subs = #sub & {#p: rd.a}
    for sn, sd in subs.dict {
      dict: "\(rn)_\(sn)": sd
    }
  }
  for rn, rd in $p.d.r if rd.a.k == "simple" {
    dict: "\(rn)": rd
  }
}

// Data
_test: {
  c1: #AC
  c1: d: {
    n: "c1"
  }

  s3: #AS
  s3: d: {
    n: "s3"
    r: r1: a:c1
  }

  s2: #AS
  s2: d: {
    n: "s2"
    r: {
      r1: a: c1
      r2: a: s3
    }
  }

  s1: #AS
  s1: d: {
    n: "s1"
    r: {
      r1: a: c1
      r2: a: s2
    }
  }
}

//  Test

out: #sub & {#p: _test.s1}
</pre>

I ran `cue export test.cue`

### What did you expect to see?
<pre>
{
    "out": {
        "dict": {
            "r1": {
                "a": {
                    "k": "simple",
                    "d": {
                        "n": "c1"
                    }
                }
            },
            "r2_r1": {
                "a": {
                    "k": "simple",
                    "d": {
                        "n": "c1"
                    }
                }
            },
            "r2_r2_r1": {
                "a": {
                    "k": "simple",
                    "d": {
                        "n": "c1"
                    }
                }
            }
        }
    }
}
</pre>

### What did you see instead?
<pre>
structural cycle:
    ./test.cue:34:16
</pre>

If I reduce in one level the structure I pass to `#sub` then the error is avoided. E.g., if the test is
`out: #pre & {#p: _test.s2}`, then I get the expected result for this case:

<pre>
{
    "out": {
        "dict": {
            "r1": {
                "a": {
                    "k": "simple",
                    "d": {
                        "n": "c1"
                    }
                }
            },
            "r2_r1": {
                "a": {
                    "k": "simple",
                    "d": {
                        "n": "c1"
                    }
                }
            }
        }
    }
}
</pre>

I tried other variations of a two-level struct, and the error is avoided in those cases, but not when the structure is more than two level deep.

**Comment by cueckoo** [comment]
_Original reply by @kumorid in https://github.com/cuelang/cue/issues/990#issuecomment-846386720_

Well,

   apparently the error appears also when the original structure is altered, as in the example presented above.

   If the structure is preserved (along the path of the recursion), the error does not appear.

   For instance, the traversal function below preserves the structure it is passed (except for the hidden field `_in`):

<pre>
#traversal: {
  _in: #AS
  
  let $in = _in.d

  k: "complex"
  d: n: $in.n
  for rn, v in $in.r {
    if v.a.k == "simple" {
      d: r: "\(rn)": v
    }

    if v.a.k == "complex" {
      d: r: "\(rn)": v & {
        a: #traversal & {_in: v.a}
      }
    }
  }
}
</pre>

Running `cue export` as in the original post on `out: #traversal & {_in: _test.s1}`computes the expected value with no cycle errors.

HOWEVER, removing unification with `v` in the recursive step within the code above (second if statement), as in,

<pre>
#traversal: {
  _in: #AS
  
  let $in = _in.d

  k: "complex"
  d: n: $in.n
  for rn, v in $in.r {
    if v.a.k == "simple" {
      d: r: "\(rn)": v
    }

    if v.a.k == "complex" {
      d: r: "\(rn)": {
        a: #traversal & {_in: v.a}
      }
    }
  }
}
</pre>

Makes a cycle error appear again:

<pre>
out.d.r.r2.a.d.r.r2.a: structural cycle
</pre>

where such unification, to my eyes, introduces nothing! 

All in all, we have not found a way to reliably perform recursive traversals within CUE above 0.2.2.

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/990#issuecomment-846397542_

Part of the problem here is that the code iterates over the same struct in which it is inserting new fields. It is technically not entirely the same struct (as it iterates over a copy). 

This is an area where arguably CUE should be more permissive. At the moment it is not, though.

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/990#issuecomment-846397724_

FTR, here is a simplified reproducer:
```
#AC: {
  k: "simple"
  n: string
}

#AS: {
  k: "complex"
  r: [string]: #AC | #AS
}

// Problematic traversal.
#sub: {
  #p: #AS
  dict: [string]: _

  for rn, rd in #p.r {
    let subs = #sub & {#p: rd}
    for sn, sd in subs.dict {
      dict: "\(rn)_\(sn)": sd
    }
  }
}

// Data
_test: {
  s3: r: r1: n: "c1"
  s2: r: r2: s3
  s1: r: r2: s2
}

//  Test

out: #sub & {#p: _test.s1}
```

**Comment by cueckoo** [comment]
_Original reply by @kumorid in https://github.com/cuelang/cue/issues/990#issuecomment-846413817_

> Part of the problem here is that the code iterates over the same struct in which it is inserting new fields. It is technically not entirely the same struct (as it iterates over a copy).
> 

Yes, as you say, it is not the same struct. 

In essence, this is a function that takes a struct, and computes over it.
The result is another struct. If there are no bounds on the depth of the struct it operates on, one needs to use a 
recursive approach to perform the transformation. In the versions of CUE above 0.2.2, this seems to raise 
cycle errors except in  narrow cases (like the one I show in my second post above)

> This is an area where arguably CUE should be more permissive. At the moment it is not, though.

To me this is incorrect behavior, is it not? 
Specially in the second case, where the behavior changes by eliminating a unification which should introduce nothing new.
That was quite baffling to me.

At the end of the day, we are talking about a recursive traversal of a structure of unknown depth, 
building a result as the traversal goes on.
From the docs, we understood this was a legitimate case of usage. We certainly have acted on that belief, although
the consequence right now is that we are stuck with version 0.2.2 of CUE, which apparently allowed such recursive traversals (it had many other problems that are gone in the latest releases, though).

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/990#issuecomment-846421752_

> Specially in the second case, where the behavior changes by eliminating a unification which should introduce nothing new.

I understand that intuition, but that doesn't mean it should be allowed. The cycle detection sometimes skips the first recursion, knowing that it is fine to detect the recursion further down, from a correctness point of view.

In this case, one could argue it should probably be allowed, though, indeed, as you mentioned.

One temporary solution may be to allow configuring an allotted recursion, allowing users, for instance, to do a total of a 1k or 1M recursions. Let me know if that would solve a pressing need.

**Comment by cueckoo** [comment]
_Original reply by @kumorid in https://github.com/cuelang/cue/issues/990#issuecomment-846509505_

Hi @mpvl , 

Thanks for the suggestion. 

Yes, I think being able to configure the number of recursions would allow us to make progress . 
The scenarios we contemplate theoretically allow arbitrary depth, but realistically, in the short term, we can cap that depth for a while.

**Comment by cueckoo** [comment]
_Original reply by @kumorid in https://github.com/cuelang/cue/issues/990#issuecomment-848835452_

Following the suggestion about limiting recursion depth, a solution can be found within CUE itself to transform a recursive function into an ordered dictionary of basically identical functions where each function, when ready to perform the recursive call, calls the next function instead.


<pre>
#RecurseN: {
  // Maximum depth of recursion
  _depth: uint
  _funcFactory: {_next: _, func: _}


  for k, it in _depth*["iter"] {
    _Funcs: "\(it)\(k)": (_funcFactory & {_next: _Funcs["\(it)\(k+1)"]}).func
  }
  _Funcs: "iter\(_depth)": null

  // Export the function at position 0
  Func: Funcs.iter0
}
</pre>

It is straightforward to obtain a function factory out of a recursive function: include all its code within the `func` output of the factory, and, instead of making the recursive call, call the `_next` parameter of the factory.

E.g., a simple depth calculator would be

<pre>
import "list"

#DepthF: {
  _next: _ 
  
  func: {
    #Value: string|int|bool|null
    _in: _ 
    out: {
      if (_in & #Value) != _|_ {
        0
      }
      if (_in & #Value) == _|_ {
        list.Max([for k,v in _in {(_next & {_in: v}).out}]) + 1
      }
    }
  }
}
</pre>

From which we could get a function that computes the depth of structures.
For instance, to compute depths up to a maximum of 10

<pre>
#depth: (#RecurseN & {_depth: 11, _funcFactory: #DepthF}).Func
</pre>

The performance hit does not seem too bad for building the dictionary of functions.

**Comment by PilchardFriendly** [comment]
This factory function was very useful - I just used this with the addition of a Recur alias, so I could traverse from multiple cases: 

```
for k, v in _in {
                let vsuite = v.#suite
                let vassert = v."assert/out"
                let Recur = (_next & { _in: v}).out
                if (vsuite != _|_) {"\(k)": Recur}
                if (vsuite == _|_) {
                    if( vassert != _|_){
                        "\(k)": vassert.message
                    }
                    if( vassert == _|_){
                        if (( v & {}) != _|_) {"\(k)": Recur}
                        if ((v & string) != _|_) {"\(k)": "passed"}
                    }
                }
            }
```

**Comment by verdverm** [comment]
I got inspired and packaged up the pattern here: https://github.com/hofstadter-io/cuetils

updates:

- simplified `RecurseN`, use an embedding to avoid `Func` setting / getting
- can set `#maxiter` on a case-by-case basis
- depth seemed to be off by one in certain cases, also supports lists

refs:
- https://github.com/hofstadter-io/cuetils/blob/main/recurse/recurse.cue
- https://github.com/hofstadter-io/cuetils/blob/main/structural/depth.cue

**Comment by WaywardWizard** [comment]
Did this get resolved? Im having trouble with recursion in 0.12. 

This is my example - i think it could fall under the umbrella of this issue as here im trying to use the function pattern recursively. Provided the input is not infinitely nested then termination will happen. Its true that there is a structural or self referential cycle here though. But cue can resolve some cycles

```
// return list of lists, each element is a path to leaf node
// keys are always strings, intermediate values are always structs
// element production depends on leaf values (a/b/c is the treepath);
//      string(ifyable): a/b/c/d/leaf
//      list: a/b/c/list[0],a/b/c/list[1],...
//          here list elements are strings
//      struct: not a leaf
#listify: {
    in: _
    if (((in & {...}) == _|_) && ((in & [...]) == _|_)) {out: [["\(in)"]]}
    if ((in & {...}) != _|_) {
        out: [
            for u, v in in
            let lv = (#listify & {"in": v}).out
            for l in lv {[u] + l},
        ]
    }
    if ((in & [...]) != _|_) {out: list.Concat([for u in in {(#listify & {in: u}).out}])}
}
```

This yields a structural cycle error...

**Comment by myitcv** [comment]
@WaywardWizard - please can you raise a new issue if you are still seeing problems?
